discovery_engine/lib/src/domain/feed_manager.dart

```diff
@@ -163,6 +164,41 @@ class FeedManager {
     await _engine.setTrustedSources(history, trustedSources);
   }
 
+  /// Override current trusted and excluded sources with the new ones provided
+  /// by the client.
+  Future<EngineEvent> setSources(
+    Set<Source> trustedSources,
+    Set<Source> excludedSources,
+  ) async {
+    final trustedFilters = trustedSources.map(SourcePreference.trusted);
+    final excludedFilters = excludedSources.map(SourcePreference.excluded);
+    final filters = {...trustedFilters, ...excludedFilters};
+    final entries = <String, SourcePreference>{};
+    final duplicates = <Source>{};
+
+    for (final filter in filters) {
+      final entry = entries.putIfAbsent(filter.source.value, () => filter);
+      if (entry.mode != filter.mode) {
+        duplicates.add(filter.source);
+      }
+    }
+
+    if (duplicates.isNotEmpty) {
+      return EngineEvent.setSourcesRequestFailed(duplicates);
+    }
+
+    final history = await _docRepo.fetchHistory();
+    await _sourcePreferenceRepository.clear();
+    await _sourcePreferenceRepository.saveAll(entries);
+    await _engine.setTrustedSources(history, trustedSources);
+    await _engine.setExcludedSources(history, excludedSources);
```

Comment:
📌 The solution is better than the current one, but I think we can make it a little bit better (but not in this pr because it probably require some more changes). 
two things that concern me:
- when we call `setTrustedSources` and `setExcludedSources` we still update the stack two times
- we call `setTrustedSources` and `setExcludedSources` (and update the stack) even though only one or none (depends on whether blue only calls it when the sources have changed) of them has changed

Reply:
Good point 👍 I've added a value equality check for sets based on implementation found in flutter's sdk, so now we call the engine only when the current set and the new one are different (added a test too). Please check if that covers the second case.

Reply:
> Please check if that covers the second case.

Nice! yes it does 👍